enhancements/machine-config/install-time-support-for-image-mode-on-openshift.md

```diff
@@ -0,0 +1,342 @@
+- Maintain OpenShift's single-click upgrade experience maintaining customizations across upgrades
+- Ensure control plane and worker nodes can all use custom OS images
+- Provide a declarative approach to OS customization through MachineOSConfig CRD
+
+### Non-Goals
+
+- Replacing the existing ignition-based configuration system entirely
+- Supporting arbitrary OS distributions beyond RHCOS-based images
+- Providing external CI/CD integration for image builds (builds occur on-cluster)
+
+## Proposal
+
+### Workflow Description
+
+The install time Image Mode on OpenShift workflow integrates with the existing OpenShift installation process in the following steps.
+
+#### Step 1: Installation Configuration
+
+1. Cluster admin runs the installer (openshift-install) to generate install time manifests
+2. Cluster admin specifies "on-cluster" image mode enablement in `install-config.yaml`
+3. Cluster admin provides `MachineOSConfig` (MOSC) yamls for master and worker MachineConfigPools
+4. Cluster admin provides/creates secret(s) needed to push and pull built images to and from the external registry
+5. Installer runs the bootkube service to combine all the manifests in the install directory to create a massive ignition file
+
+#### Step 2: Bootstrap Phase
+
+1. Bootstrap Machine Config Operator(MCO) parses the massive ignition file created by the bootkube service during the installer phase
+2. Bootstrap MCO/MCC parses the ignition file to generate the necessary MachineConfigs
+3. Bootstrap MCO/MCC pods establishes connection with external image registry (for this enhancement, running a registry on the bootstrap node is a non-goal)
+4. Once the rendered-MC is generated, bootstrap MCO/MCC starts the "on-cluster" image build process by creating a pod to build the custom OS image for the master MCP and push the built image to the external registry
+5. Bootstrap MCO/MCC creates a MachineConfig to override the `osImageURL` to point to the newly built image for the master pool
+6. Bootstrap node creates the `control plane` nodes (master pool) with the stock AMI
+7. Bootstrap MCO runs `bootc switch` or `rpm-ostree rebase` to boot the master nodes into the custom built OS image refered to by the `osImageURL` value
+
+#### Step 3: Worker Node Deployment
+
+1. Wait for the rendered-MC for the worker pool to be generated
+2. Wait for all the configs defined during install time to be applied to the cluster - the worker MOSC should be part of this
+3. Control plane nodes parse the worker MOSC configuration to start the "on-cluster" build process similar to how it is currently done for post-install image mode
+4. The control plane node which is running the MachineOSBuilder pod establishes connection with external image registry
+5. The control plane node which is running the MachineOSBuilder pod pushes the built image to the registry
+6. MCO/MCC creates a MachineConfig to override the `osImageURL` for the worker pool to point to the newly built image
```

Comment:
Same as above, isn't this conflicting with how regular OCL works? Shouldn't' we be just setting the desiredImage? (apologies if I'm confusing the workflows here)

Reply:
Similar question as above - I believe the workflow for overriding the image would be the same for master and worker nodes.